src/Symfony/Component/Security/Http/EventListener/UserCheckerListener.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\User\UserCheckerInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractPreAuthenticatedAuthenticator;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class UserCheckerListener implements EventSubscriberInterface
+{
+    private $userChecker;
+
+    public function __construct(UserCheckerInterface $userChecker)
+    {
+        $this->userChecker = $userChecker;
+    }
+
+    public function preCredentialsVerification(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        if (null === $event->getUser() || $event->getAuthenticator() instanceof AbstractPreAuthenticatedAuthenticator) {
```

Comment:
The `AbstractPreAuthenticatedAuthenticator` seems a bit arbitrary. Perhaps we should add some `shouldEnforceUserChecker()` method to `AuthenticatorInterface` (or some other interface) then implement that to return `true` in `AbstractAuthenticator` (so that it effectively defaults to true). Then, in `AbstractPreAuthenticatedAuthenticator` we will return `false` (or not implement the interface, if you think adding a new interface is a better approach).